docs: use regex for Vite server.cors to support additional hostnames (#313)
ddev/ddev.com: src/content/blog/working-with-vite-in-ddev.md

     // Adjust Vites dev server to work with DDEV
     // https://vitejs.dev/config/server-options.html
     server: {
       // Respond to all network requests
       host: "0.0.0.0",
       port: port,
       strictPort: true,
       // Defines the origin of the generated asset URLs during development,
       // this must be set to the Vite devserver URL and selected port

>        origin: `${process.env.DDEV_PRIMARY_URL}:5173`,

Reviewer: ```suggestion
       origin: `${process.env.DDEV_PRIMARY_URL.replace(/:\d+$/, '')}:5173`,
```

       // Configure CORS securely for the Vite dev server to allow requests
       // from *.ddev.site domains, supports additional hostnames (via regex)
       cors: {

>          origin: /https?:\/\/([A-Za-z0-9\-\.]+)?(\.site)(?::\d+)?$/,

Reviewer 2: Thanks for writing the post 🙏 

I think this regex should be
```suggestion
         origin: /https?:\/\/([A-Za-z0-9\-\.]+)?(\.ddev\.site)(?::\d+)?$/,
```
Otherwise, an attacker can serve their malicious script on `*.site` and the regex in the PR will also that script to fetch the content.

Author: Hi @sapphi-red, thanks very much that you also review here 👍 

(we had a discussion here https://github.com/vitejs/vite/issues/19345, sapphi-red is vitejs core team member)

Oh, I wasn't aware that .site is also an official TLD. Thought it was only local.

I'll change it.

Author: Changed ✅

     // Adjust Vites dev server to work with DDEV
     // https://vitejs.dev/config/server-options.html
     server: {
       // respond to all network requests:
       // Respond to all network requests
       host: "0.0.0.0",
       port: port,

>        strictPort: true,

Reviewer: `strictPort` still should be here, right?

>        // Defines the origin of the generated asset URLs during development, this must be set to the

Reviewer: ```suggestion
       strictPort: true,
       // Defines the origin of the generated asset URLs during development, this must be set to the
```

Author: thx for re-checking /proofreading - very helpful, resolved ✅

       strictPort: true,
       // Defines the origin of the generated asset URLs during development
       origin: origin,
       // Configure CORS for the devserver (security)
       cors: {

>          origin: /https?:\/\/([A-Za-z0-9\-\.]+)?(localhost|\.site)(?::\d+)?$/,

Reviewer: Can we use `process.env.DDEV_TLD` here?

People can use different `project_tld` instead of `ddev.site`.

Author: I need to check https://stackoverflow.com/questions/494035/how-do-you-use-a-variable-in-a-regular-expression 🤔

(Might be maybe simpler to add a section with a note about custom tlds and that users need to change it)

  // Adjust Vites dev server for DDEV
  // https://vitejs.dev/config/server-options.html
  server: {
    // Respond to all network requests
    host: "0.0.0.0",
    port: 5173,
    strictPort: true,
    // Defines the origin of the generated asset URLs during development,
    // this must be set to the Vite devserver URL and selected port

>     origin: `${process.env.DDEV_PRIMARY_URL}:5173`,

Reviewer: ```suggestion
    origin: `${process.env.DDEV_PRIMARY_URL.replace(/:\d+$/, '')}:5173`,
```

    // can be configured to use another port (via `router_https_port`), the output can also be
    // "https://test-vite.ddev.site:1234". Therefore we need to strip a port number like ":1234"
    // before adding Vites port to achieve the desired output of "https://test-vite.ddev.site:5173".
    origin: `${process.env.DDEV_PRIMARY_URL.replace(/:\d+$/, "")}:5173`,
    // Configure CORS securely for the Vite dev server to allow requests from *.ddev.site domains,
    // supports additional hostnames (via regex). If you use another `project_tld`, adjust this.
    cors: {
      origin: /https?:\/\/([A-Za-z0-9\-\.]+)?(\.ddev\.site)(?::\d+)?$/,
    },
  },
})
```

Technical explanation:

- We need to define an entry file to let vite know where to start, this is done in `rollupOptions`.
- Also we need to tell Vite to respond to all network request, not only the ones addressed internally to http://localhost:5173. This is done via `host: '0.0.0.0'` and is important for making it work with DDEV / Docker.
- The strict port setting is also necessary, because we only exposed port 5173. Without `strictPort: true`, Vite will use other ports like 5174 or 5175 if port 5173 is already occupied.

> - Another important part is to let Vite know from where to load Vite-controlled assets like images referenced in CSS. This is done via `server.origin`. <br> _(DDEV automatically provides environment variables via the regular `process.env` variable. The variable `process.env.DDEV_PRIMARY_URL` will have the value `https://test-vite.ddev.site` in our demo project. We use it to set the correct `server.origin` dynamically.)_ DDEV can be configured to use other ports than 80/443, or will auto-select ports if these are occupied already. Therefore we need to support both possible outputs of `process.env.DDEV_PRIMARY_URL`: "https://test-vite.ddev.site" or somehting like "https://test-vite.ddev.site:1234". If there is a custom port, we need to strip it before we add the Vite port.

Reviewer: ```suggestion
- Another important part is to let Vite know from where to load Vite-controlled assets like images referenced in CSS. This is done via `server.origin`. <br> _(DDEV automatically provides environment variables via the regular `process.env` variable. The variable `process.env.DDEV_PRIMARY_URL` will have the value `https://test-vite.ddev.site` in our demo project. We use it to set the correct `server.origin` dynamically.)_ <br> DDEV can be configured to use other ports than 80/443, or will auto-select ports if these are occupied already. Therefore we need to support both possible outputs of `process.env.DDEV_PRIMARY_URL`: "https://test-vite.ddev.site" or something like "https://test-vite.ddev.site:1234". If there is a custom port, we need to strip it before we add the Vite port.
```

- Last but not least, we need to configure secure CORS headers. In earlier versions of Vite, this was set to `cors: true` by default and allowed all origins to fetch scripts from the devserver. After [security advisory GHSA-vg6x-rcgg-rjx6](https://github.com/vitejs/vite/security/advisories/GHSA-vg6x-rcgg-rjx6), the default setting was changed and we need to add the allowance of DDEV local domains explicitly. The regular expression supports domains like `https://test-vite.ddev.site` (or something like `https://test-vite.ddev.site:1234` when you use another HTTPS port). See [server.cors](https://vite.dev/config/server-options#server-cors) in the Vite docs for more information.

> If your site runs on another [top-level-domains (project_ltd)](https://ddev.readthedocs.io/en/stable/users/configuration/config/#project_tld) rather than `.ddev.site`, you edit the regex yourself or use this automagic snippet:

Reviewer: ```suggestion
If your site runs on another [top-level-domains (project_ltd)](https://ddev.readthedocs.io/en/stable/users/configuration/config/#project_tld) rather than `.ddev.site`, you edit the regular expression yourself or use this automagic snippet:
```

Afterwards, you just need to change the `vite.config.js` like this:

```js
import { defineConfig } from "vite"
import laravel from "laravel-vite-plugin"

export default defineConfig({
  plugins: [
    laravel({
      input: ["resources/css/app.css", "resources/js/app.js"],
      refresh: true,
    }),
  ],
  server: {
    // Respond to all network requests
    host: "0.0.0.0",
    port: 5173,
    strictPort: true,
    // Defines the origin of the generated asset URLs during development,
    // this will also be used for the public/hot file (Vite devser + port)

>     origin: `${process.env.DDEV_PRIMARY_URL}:5173`,

Reviewer: ```suggestion
    origin: `${process.env.DDEV_PRIMARY_URL.replace(/:\d+$/, '')}:5173`,
```
